slidechain/export.go

```diff
@@ -276,67 +342,81 @@ func BuildExportTx(ctx context.Context, asset xdr.Asset, amount, inputAmt int64,
 		Temp     string `json:"temp"`
 		Seqnum   int64  `json:"seqnum"`
 		Exporter string `json:"exporter"`
+		Amount   int64  `json:"amount"`
+		Anchor   []byte `json:"anchor"`
+		Pubkey   []byte `json:"pubkey"`
 	}{
 		assetXDR,
 		temp,
 		int64(seqnum),
 		kp.Address(),
+		inputAmt,
+		anchor,
+		pubkey,
 	}
-	refdata, err := json.Marshal(ref)
+	inputRefdata, err := json.Marshal(ref)
 	if err != nil {
 		return nil, errors.Wrap(err, "marshaling reference data")
 	}
-	tpl := txbuilder.NewTemplate(time.Now().Add(time.Minute), nil)
-	tpl.AddInput(1, [][]byte{prv}, nil, []ed25519.PublicKey{pubkey}, inputAmt, assetID, anchor, nil, 1)
-	tpl.AddRetirement(int64(amount), assetID, refdata)
-	if inputAmt > amount {
-		tpl.AddOutput(1, []ed25519.PublicKey{pubkey}, inputAmt-amount, assetID, nil, nil)
+	inputRefdataHex := i10rjson.HexBytes(inputRefdata)
+
+	ref.Amount = amount
+	// We first split off the difference between inputAmt and amt.
+	// Then, we split off the zero-value for finalize, creating the retire anchor.
+	retireAnchor1 := txvm.VMHash("Split2", anchor)
+	retireAnchor := txvm.VMHash("Split1", retireAnchor1[:])
+	ref.Anchor = retireAnchor[:]
+	retireRefdata, err := json.Marshal(ref)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshaling reference data")
 	}
-	err = tpl.Sign(ctx, func(_ context.Context, msg []byte, prv []byte, path [][]byte) ([]byte, error) {
-		return ed25519.Sign(prv, msg), nil
+	retireRefdataHex := i10rjson.HexBytes(retireRefdata)
+	b := new(txvmutil.Builder)
+	b.PushdataBytes(inputRefdataHex)                                                                                     // con stack: json
+	b.Op(op.Put)                                                                                                         // arg stack: json
+	standard.SpendMultisig(b, 1, []ed25519.PublicKey{pubkey}, inputAmt, assetID, anchor, standard.PayToMultisigSeed1[:]) // arg stack: inputval, sigcheck
+	b.Op(op.Get).Op(op.Get)                                                                                              // con stack: sigcheck, inputval
+	b.PushdataInt64(amount).Op(op.Split)                                                                                 // con stack: sigcheck, changeval, retireval
+	b.PushdataInt64(1).Op(op.Roll).Op(op.Drop)                                                                           // con stack: sigcheck, retireval
```

Comment:
You can only `drop` a value if it's zero! Which is only true if `amount == inputAmt`. If it's non-zero, you need to pay it as change (e.g. using `standard.PayToMultisigProg1`) to the owner of `pubkey`.